[10.0] barcodes_generator_*: add image generation wizard (+ usability improvements)

alexis-via opened the pull request:
Don't use parenthesis in button labels, use help message to give more info.
Hide the button "Generate Barcode" when a barcode is already present.

alexis-via committed:
barcodes_generator_product: usability and string improvements

Don't use parenthesis in button labels, use help message to give more info.
Hide the button "Generate Barcode" when a barcode is already present.

legalsylvain commented on barcodes_generator_product/views/view_product_product.xml:
```diff
@@ -21,21 +21,23 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+                    help="Generate Barcode Base from the sequence configured on the Barcode Rule"
+                    attrs="{'invisible': ['|',
+                        ('generate_type', '!=', 'sequence'),
+                        ('barcode_base', '!=', 0)]}"
```
LGTM.

alexis-via committed:
barcodes_generator_abstract: Hide button "Generate Sequence" when a sequence is already set

legalsylvain commented on barcodes_generator_product/views/view_product_product.xml:
```diff
@@ -21,21 +21,23 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
                             ('barcode_rule_id', '=', False),
-                            ('barcode_base', '=', 0)]}"
+                            ('barcode_base', '=', 0),
+                            ('barcode', '!=', False)]}"
```
Not convinced for this point.  (just the ('barcode', '!=', False) condition.

This will prevent users to regenerate barcode, that is usefull  if you change the pattern of your barcode, changing prefix from 021... to 022.... for exemple.

alexis-via replied:
Right, I didn't have this scenario in mind (not sure if changing barcodes on existing product is a scenario that is supposed to happen, but...). In this case, they would have to empty the barcode field and then the button will be displayed again.

By the way, I forgot to say in the comments that I also removed the readonly attribute on the barcode field when generated from sequence so that users can remove a barcode from a product (this happens when they re-use an existing barcode for another product)

legalsylvain replied:
OK. Well, the problem I see in this configuration is the possiblity to write a bad barcode associated to a product with barcode_rule_id defined.

A medium proposal could be : 
- hide button generate Barcode if barcode is defined. (what you already did)
- let barcode readonly (to secure data). (by the way, should be readonly if barcode_rule_id != False).
- Add an extra button "Clear barcode" that set barcode to False, displayed if barcode is defined and barcode_rule_id is defined.

what do you think about that proposal ? More secure, and allow to change barcodes.

alexis-via replied:
I find it a bit heavy... but I can understand that when you have "untrusted" users for which you want to lock down things and block mistakes as much as possible. So I just implemented it your way.

legalsylvain replied:
Thanks a lot. :+1:

legalsylvain commented on barcodes_generator_product/views/view_product_template.xml:
```diff
@@ -21,21 +21,23 @@ License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
                             ('barcode_rule_id', '=', False),
-                            ('barcode_base', '=', 0)]}"
+                            ('barcode_base', '=', 0),
+                            ('barcode', '!=', False)]}"
```
same here.

legalsylvain reviewed (changes requested):
one question the rest LGTM.

thanks for improving this module.

alexis-via commented:
@legalsylvain In the README.rst of barcodes_generator_abstract, in the installation section, there is:

```
'viivakoodi' is a more active for of 'pyBarcode'.
```

Did you mean "a more active fork" ? If yes, i'll fix this in this PR.

legalsylvain commented:
> Did you mean "a more active fork" ? If yes, i'll fix this in this PR.

indeed, thanks !

alexis-via committed:
barcodes_generator_abstract: improve README

alexis-via committed:
barcodes_generator_abstract: add barcode image generation wizard

barcodes_generator_partner and barcodes_generator_product: use the barcode image generation wizard defined in barcodes_generator_abstract

legalsylvain commented:
barcode image generation is a good improvment. Thanks.

rousseldenis commented on barcodes_generator_abstract/README.rst:
```diff
@@ -6,34 +6,37 @@
-This module use an extra python library named 'viivakoodi' you should install
-to make barcode generation works properly. 'viivakoodi' is a more active for of
-'pyBarcode'.
+This module use an extra python library named *viivakoodi* you should install
```
@alexis-via You should declare it in external_dependecies manifest part

rousseldenis replied:
Ok, did'nt see https://github.com/OCA/stock-logistics-barcode/pull/101/

rousseldenis commented on barcodes_generator_abstract/wizard/barcode_image_generate.py:
```diff
@@ -0,0 +1,123 @@
+            res.append((barcode_format, barcode_format))
+        return res
+
+    state = fields.Selection([
```
Maybe a help part to describe briefly each option?

rousseldenis reviewed (changes requested):
Code review. Minor comments.
LGTM

rousseldenis commented:
@legalsylvain Can you review last changes?

legalsylvain commented:
@rousseldenis,
https://github.com/OCA/stock-logistics-barcode/pull/104#discussion_r136033225 is not answered. Or did I missed something ? 

regards.

alexis-via committed:
Restore readonly property on product barcode field when barcode_rule_id is set

Add button "Empty barcode" when the barcode is in readonly

legalsylvain reviewed (approved):
Thanks for all the improvment.

code review. no test.

legalsylvain commented on barcodes_generator_abstract/wizard/barcode_image_generate.py:
```diff
@@ -0,0 +1,123 @@
+        help="Vertical size of the barcode")
+    module_width = fields.Float(
+        default=0.2, string='Barcode Width',
+        help="It doesn't seem to have any effect in EAN13...")
```
Sorry, I'm rechecking this PR. I don't understand this helper. do you mean this field is useless : 
- allways ? 
- for ean13 ? 

in all cases, i don't think that this helper will help end user. Maybe a comment should be more usefull. (and maybe make this field hidden, if it is useless. don't you think ?
